### PR TITLE
feat: added support to define custom widget in screen

### LINKED ISSUE #529: Support defining custom widget in screen

### PR DESCRIPTION
**Pull request by sagardspeed2**
## Describe your changes
Added support to define custom widget in screen

## Issue ticket number and link

Closes #529 

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] I have added tests
- [x] I have added a changeset `pnpm changeset add`
- [x] I have added example usage in the kitchen sink app

### CODE REVIEW

File: packages/runtime/src/runtime/screen.tsx [hunk 1]
```diff
@@ -37,6 +40,28 @@ export const EnsembleScreen: React.FC<EnsembleScreenProps> = ({
...
+      return;
+    }
+
+    // load screen custom widgets
+    screen.customWidgets.forEach((customWidget) => {
+      WidgetRegistry.register(
+        customWidget.name,
+        createCustomWidget(customWidget),
+      );
+    });
```
Comment (evshi) [inline packages/runtime/src/runtime/screen.tsx:56]: This does not scope the custom widget to the current screen.

**Review by evshi — CHANGES_REQUESTED** [review]

File: packages/runtime/src/registry.tsx [hunk 2]
```diff
@@ -1,20 +1,36 @@
 import { Alert } from "antd";
 import type { ReactElement } from "react";
 
+type WidgetComponent<T> = React.FC<T>;
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const mainRegistry: Record<string, WidgetComponent<any> | undefined> = {};
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-const registry: Record<string, React.FC<any> | undefined> = {};
+const backupRegistry: Record<string, WidgetComponent<any> | undefined> = {};
```
Comment (evshi) [inline packages/runtime/src/registry.tsx:9]: What is the purpose of having two registries?

File: packages/runtime/src/registry.tsx [hunk 3]
```diff
@@ -1,20 +1,36 @@
...
-  find: (name: string): React.FC | ReactElement => {
-    const Widget = registry[name];
+  find: (name: string): WidgetComponent<any> | ReactElement => {
+    const Widget = mainRegistry[name];
     if (!Widget) {
       return <UnknownWidget missingName={name} />;
     }
     return Widget;
   },
+  unMount: (name: string): void => {
```
Comment (evshi) [inline packages/runtime/src/registry.tsx:26]: Rename to `unregister`. Functions should be named symmetrically.

File: packages/runtime/src/registry.tsx (hunk 2 above)
Comment (sagardspeed2) [reply packages/runtime/src/registry.tsx:9]: When user created any custom widget with the same name of our in-build widgets, that time we can store our in-built widget into backupRegistry and load the custom widget into mainRegistry and serve the widget, and when screen gets destroyed, we remove that custom widget from mainRegistry and pick our in-built widget from the backupRegistry and load into mainRegistry

**Review by evshi — COMMENTED** [review]
You don't need an entirely separate map - you can just memoize the old implementation and restore it on unmount.

**Comment by sagardspeed2** [comment]
> You don't need an entirely separate map - you can just memoize the old implementation and restore it on unmount.

This is just one object that useful to memoize the old implementation by storing that implementation temporary and then when screen unregister, we again remove the old implementation from the object.

File: packages/runtime/src/registry.tsx [hunk 4]
```diff
@@ -1,20 +1,25 @@
 import { Alert } from "antd";
 import type { ReactElement } from "react";
 
+type WidgetComponent<T> = React.FC<T>;
+
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-const registry: Record<string, React.FC<any> | undefined> = {};
+export const registry: { [key: string]: WidgetComponent<any> | undefined } = {};
```
Comment (evshi) [inline packages/runtime/src/registry.tsx:7]: Don't export the registry, it is supposed to be private to this module.

File: packages/runtime/src/runtime/screen.tsx [hunk 5]
```diff
@@ -26,6 +28,7 @@ export const EnsembleScreen: React.FC<EnsembleScreenProps> = ({
 }) => {
   // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
   const { state, search, pathname } = useLocation();
+  const [isLoaded, setIsLoaded] = useState<boolean>(false);
```
Comment (evshi) [inline packages/runtime/src/runtime/screen.tsx:31]: Can we rename to isInitialized?

File: packages/runtime/src/runtime/screen.tsx [hunk 6]
```diff
@@ -37,6 +40,51 @@ export const EnsembleScreen: React.FC<EnsembleScreenProps> = ({
...
+    }
+
+    // initial widget values store
+    const initialWidgetValues: {
+      [key: string]: WidgetComponent<any>;
+    } = {};
+
+    // load screen custom widgets
+    screen.customWidgets?.forEach((customWidget) => {
+      const originalImplementation = findWidget(customWidget.name);
```
Comment (evshi) [inline packages/runtime/src/runtime/screen.tsx:57]: Keep this on WidgetRegistry object. Name it `findOrNull`

**Review by evshi — APPROVED** [review]

**Pull request merged**